pkg/stacks/unpack.go

```diff
@@ -17,320 +17,439 @@ limitations under the License.
 	yamlSeparator           = "\n---\n"
+
+	// Stack annotation constants
+	annotationStackIcon           = "stacks.crossplane.io/icon"
+	annotationStackUISpec         = "stacks.crossplane.io/ui-spec"
+	annotationStackTitle          = "stacks.crossplane.io/stack-title"
+	annotationGroupTitle          = "stacks.crossplane.io/group-title"
+	annotationGroupCategory       = "stacks.crossplane.io/group-category"
+	annotationGroupDescription    = "stacks.crossplane.io/group-description"
+	annotationResourceTitle       = "stacks.crossplane.io/resource-title"
+	annotationResourceTitlePlural = "stacks.crossplane.io/resource-title-plural"
+	annotationResourceCategory    = "stacks.crossplane.io/resource-category"
+	annotationResourceDescription = "stacks.crossplane.io/resource-description"
+	annotationKubernetesManagedBy = "app.kubernetes.io/managed-by"
 )
 
 var (
 	log = logging.Logger.WithName("stacks")
 )
 
-// Unpack unpacks the stack contents from the given directory.
-func Unpack(contentDir string) error {
-	log.V(logging.Debug).Info("Unpacking stack", "contentDir", contentDir)
-	fs := afero.NewOsFs()
+// StackResource provides the Stack meta-data for a CRD. This is the format for resource.yaml files.
+type StackResource struct {
+	// ID refers to the CRD Kind
+	ID          string `json:"id"`
+	Title       string `json:"title"`
+	TitlePlural string `json:"title-plural"`
+	Description string `json:"description"`
+	Category    string `json:"category"`
+}
 
-	registryRoot := findRegistryRoot(fs, contentDir)
+// StackGroup provides the Stack meta-data for a resource group. This is the format for group.yaml files.
+type StackGroup struct {
+	Title       string `json:"title"`
+	Description string `json:"description"`
+	Category    string `json:"category"`
+}
 
-	content, err := doUnpack(fs, registryRoot)
-	if err != nil {
-		return err
-	}
+// StackPackager implentations can build a stack from Stack resources and emit the Yaml artifact
+type StackPackager interface {
+	SetApp(v1alpha1.AppMetadataSpec)
+	SetInstall(unstructured.Unstructured) error
+	SetRBAC(v1alpha1.PermissionsSpec)
 
-	// write content to stdout
-	_, err = os.Stdout.WriteString(content)
+	GotApp() bool
 
-	return err
+	AddGroup(string, StackGroup)
+	AddResource(string, StackResource)
+	AddIcon(string, v1alpha1.IconSpec)
+	AddUI(string, string)
+	AddCRD(string, *apiextensions.CustomResourceDefinition)
+
+	Yaml() string
 }
 
-func findRegistryRoot(fs afero.Fs, dir string) string {
-	if _, err := fs.Stat(filepath.Join(dir, registryDirName)); err == nil {
-		// the .registry subdir exists under the given dir, that must be the registry root
-		return filepath.Join(dir, registryDirName)
-	}
+// StackPackage defines the artifact structure of Stacks
+// A fully processed Stack can be thought of as a Stack CR (with its install and RBAC fields defined) and a collection of managed CRDs (annotated by their Stack resource, icon, group, and UI descriptors)
```

Comment:
nit: there are some really long comment lines in this PR (this one is over 200 chars), can you shorten some of them? they are hard to read especially in the GitHub PR review UI which is fairly narrow.

Reply:
It's looks like GH offers an ~80 character width. 80 it is. I'll also take a copy-edit pass on these comments since I've drifted from their original text.

I'm not sure why `lll` didn't flag this line.  The docs say it will only ignore generated comments over the default max-length of 80 characters.  Should we add something about this to the CONTRIBUTING.md?

Now if only we could keep the functions in under 24 lines :tv:

Reply:
[Resolved](https://github.com/crossplaneio/crossplane/compare/9f5ab3c6d421d5e1b0d722af3317569614d5132f..017c3d47090daca29f82dfd30b95578a3b3e297d)

Reply:
i've traditionally used 100 chars and maybe 120 max, 80 chars feels pretty narrow.  we should probably have a standard practice added to the contribution guide.

i'm just fine with what you've done so far here, no changes needed :)